src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/ReferencesPage/ImportedNamespacesValueProvider.cs

```diff
@@ -0,0 +1,132 @@
+                {
+                    containsProjectName = true;
+                }
+                sb.Append($"{import};");
+            }
+        }
+
+        if (!containsProjectName)
+        {
+            sb.Append($"{projectName};");
+        }
+
+        sb.Remove(sb.Length - 1, 1);
+        return sb.ToString();
+    }
+    
+    public override Task<string> OnGetUnevaluatedPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties)
+    {
+        return GetSelectedImportStringAsync();
+    }
+
+    public override Task<string> OnGetEvaluatedPropertyValueAsync(string propertyName, string evaluatedPropertyValue, IProjectProperties defaultProperties)
+    {
+        return GetSelectedImportStringAsync();
+    }
+
+    public override async Task<string?> OnSetPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string>? dimensionalConditions = null)
+    {
+        await _threadingService.SwitchToUIThread();
+        
+        string[] newImportsToSet = unevaluatedPropertyValue.Split(';');
+        // delete existing imports that aren't in unevaluatedPropertyValue
+        
+        // add imports that are in unevaluatedPropertyValue but not in _imports
+        var importsToAdd = newImportsToSet.ToHashSet();
+        importsToAdd.Remove(Path.GetFileNameWithoutExtension(_configuredProject.UnconfiguredProject.FullPath));
+
+        foreach (string import in await GetSelectedImportsAsync())
+        {
+            if (!newImportsToSet.Contains(import))
+            {
+                try
+                {
+                    await _projectAccessor.OpenProjectForWriteAsync(_configuredProject, project =>
+                    {
+                        Microsoft.Build.Evaluation.ProjectItem importProjectItem = project.GetItems("Import")
+                            .First(i => string.Equals(import, i.EvaluatedInclude, StringComparisons.ItemNames));
+
+                        if (importProjectItem.IsImported)
+                        {
+                            _specialImports.Add(import);
+                        }
+                        
+                        project.RemoveItem(importProjectItem);
+                    });
+
+                }
+                catch (Exception)
+                {
+                    // if an import comes from a targets file, or else if there's a race condition we can't remove it
+                }
+            }
+            
+            importsToAdd.Remove(import);
+        }
+
+        foreach (string importToAdd in importsToAdd)
+        {
+            if (importToAdd.Length > 0)
+            {
+                await _projectAccessor.OpenProjectXmlForWriteAsync(_configuredProject.UnconfiguredProject, project =>
+                {
+                    project.AddItem("Import", importToAdd);
+                });
+            }
+        }
```

Comment:
Getting a write lock is "expensive" so shouldn't be done in a loop. Something like this might help: 

```suggestion
        // Verify we have at least one valid import to add before acquiring a write lock.
        if (importsToAdd.Any(importToAdd => importToAdd.Length > 0))
        {
            await _projectAccessor.OpenProjectXmlForWriteAsync(_configuredProject.UnconfiguredProject, project =>
            {
                foreach (string importToAdd in importsToAdd)
                {
                    if (importToAdd.Length > 0)
                    {
                        project.AddItem("Import", importToAdd);
                    }
                }
            });
        }
```